src/lib/middleware/response-time-metrics.ts

```diff
@@ -2,21 +2,36 @@ import * as responseTime from 'response-time';
 import EventEmitter from 'events';
 import { REQUEST_TIME } from '../metric-events';
 import { IFlagResolver } from '../types/experimental';
+import { InstanceStatsService } from 'lib/services';
+import memoizee from 'memoizee';
 
 // eslint-disable-next-line @typescript-eslint/naming-convention
 const _responseTime = responseTime.default;
 
+const countAppNames = async (instanceStatsService: InstanceStatsService) => {
+    const response = await instanceStatsService.getLabeledAppCounts();
+    return response.find((c) => c.range === '7d').count;
+};
 export function responseTimeMetrics(
     eventBus: EventEmitter,
     flagResolver: IFlagResolver,
+    instanceStatsService: InstanceStatsService,
 ): any {
-    return _responseTime((req, res, time) => {
+    const appNameReportingThreshold = 100;
+    let appNameCount = memoizee(() => countAppNames(instanceStatsService), {
+        promise: true,
+        maxAge: 30000, // milliseconds
+    });
+    return _responseTime(async (req, res, time) => {
         const { statusCode } = res;
 
         const pathname = req.route ? req.baseUrl + req.route.path : '(hidden)';
 
         let appName;
-        if (flagResolver.isEnabled('responseTimeWithAppName')) {
+        if (
+            flagResolver.isEnabled('responseTimeWithAppName') &&
+            (await appNameCount()) < appNameReportingThreshold
```

Comment:
I am unsure if this does exactly what we want? I know we are deleting instance stats, but will we keep application information forever? If yes, then we should be fine.

Reply:
Based on our conversation, appNameCount means `countDistinctAppNamesFromInstances` (i.e. we're maintaining a sliding window of active instances from which we fetch the app names). 

We're caching this count to avoid recalculating on every request, but it's currently done as a write-ahead cache, so it'll likely impact p99 latency. The other option would be to implement a write-behind cache so we don't spend time going to the DB in the user thread

> I know we are deleting instance stats, but will we keep application information forever? If yes, then we should be fine.

We do keep application information forever (we're not changing that), but we're using the instance stats for this count which are deleted after some time. This allows customers who had a surge of instances registered with random application names to recover and re-gain this metric once the older instances are deleted (because this count will go down)